File: quickmail/quickmail/Core/API/OriginValidator.swift
import Foundation

nonisolated enum OriginValidator {
    static func validate(_ rawValue: String) throws -> URL {
        let value = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !value.isEmpty, var components = URLComponents(string: value) else {
            throw APIError.invalidOrigin("Enter your QuickMail server URL.")
        }

        guard components.user == nil, components.password == nil else {
            throw APIError.invalidOrigin("The server URL cannot contain credentials.")
        }
        guard components.query == nil, components.fragment == nil else {
            throw APIError.invalidOrigin("Use only the QuickMail server origin, without a query or fragment.")
        }
        guard components.path.isEmpty || components.path == "/" else {
            throw APIError.invalidOrigin("Use only the QuickMail server origin, without a path.")
        }
        guard let host = components.host, !host.isEmpty else {
            throw APIError.invalidOrigin("The server URL must include a host.")
        }

        let scheme = components.scheme?.lowercased()
        #if DEBUG
        let isLocalHTTP = scheme == "http" && ["localhost", "127.0.0.1", "::1"].contains(host.lowercased())
        #else
        let isLocalHTTP = false
        #endif
        guard scheme == "https" || isLocalHTTP else {
            throw APIError.invalidOrigin("QuickMail requires a secure HTTPS server URL.")
        }

        components.scheme = scheme
        components.host = host.lowercased()
        components.path = ""
        guard let url = components.url, url.absoluteURL == url else {
            throw APIError.invalidOrigin("The QuickMail server URL is invalid.")
        }
        return url
    }
}

File: android/app/src/main/java/dev/anuz/quickinbox/data/OriginValidator.kt
package dev.anuz.quickinbox.data

import android.net.Uri
import dev.anuz.quickinbox.BuildConfig

object OriginValidator {
    fun validate(rawValue: String): String {
        val value = rawValue.trim()
        if (value.isEmpty()) throw ApiError.InvalidOrigin("Enter your QuickInbox server URL.")

        val uri = Uri.parse(value)
        if (!uri.userInfo.isNullOrEmpty()) {
            throw ApiError.InvalidOrigin("The server URL cannot contain credentials.")
        }
        if (!uri.query.isNullOrEmpty() || !uri.fragment.isNullOrEmpty()) {
            throw ApiError.InvalidOrigin("Use only the QuickInbox server origin, without a query or fragment.")
        }
        val path = uri.path.orEmpty()
        if (path.isNotEmpty() && path != "/") {
            throw ApiError.InvalidOrigin("Use only the QuickInbox server origin, without a path.")
        }
        val host = uri.host?.lowercase().orEmpty()
        if (host.isEmpty()) throw ApiError.InvalidOrigin("The server URL must include a host.")

        val scheme = uri.scheme?.lowercase()
        val isLocalHttp = BuildConfig.DEBUG &&
            scheme == "http" &&
            host in setOf("localhost", "127.0.0.1", "::1")
        if (scheme != "https" && !isLocalHttp) {
            throw ApiError.InvalidOrigin("QuickInbox requires a secure HTTPS server URL.")
        }

        return Uri.Builder()
            .scheme(scheme)
            .encodedAuthority(host + (if (uri.port != -1) ":${uri.port}" else ""))
            .build()
            .toString()
            .trimEnd('/')
    }
}
